librubyfmt/src/format_prism.rs

```diff
@@ -235,6 +242,184 @@ fn format_statements(ps: &mut dyn ConcreteParserState, statements_node: prism::S
+                let heredoc_kind = HeredocKind::from_string(&heredoc_symbol);
+
+                ps.emit_heredoc_start(heredoc_symbol, heredoc_kind);
+                ps.emit_newline();
+            } else {
+                // Always use double quotes over single quotes/percent literals
+                if opener.is_some() {
+                    ps.emit_double_quote();
+                }
+            }
+
+            // If opener is nil, we must be in some kind of interpolated string context, which
+            // means the contents must already be appropriately escaped -- hence we default to `true` here
+            let in_escaped_context =
+                is_heredoc || opener.clone().map(|s| s.starts_with("\"")).unwrap_or(true);
+            let string_content = if in_escaped_context {
+                loc_to_string(string_node.content_loc())
+            } else {
+                crate::string_escape::single_to_double_quoted(
+                    loc_to_string(string_node.content_loc()),
+                    opener.clone().unwrap().as_str(),
+                    closer.clone().unwrap().as_str(),
+                )
+            };
+
+            ps.emit_string_content(string_content);
+            ps.wind_dumping_comments_until_offset(string_node.content_loc().end_offset());
+
+            if is_heredoc {
+                // <<~ heredocs have their closing tag indented too
+                if opener.map(|s| s.starts_with("<<~")).unwrap_or(false) {
+                    ps.emit_indent();
+                }
+
+                ps.emit_heredoc_close(
+                    loc_to_string(string_node.closing_loc().unwrap())
+                        .trim()
+                        .to_string(),
+                );
+            } else if opener.is_some() {
+                ps.emit_double_quote();
+            }
+        }),
+    );
+
+    ps.wind_dumping_comments_until_offset(string_node.location().end_offset());
+
+    if ps.at_start_of_line() {
+        ps.emit_newline();
+    }
+}
+
+fn format_interpolated_string_node(
+    ps: &mut dyn ConcreteParserState,
+    interpolated_string_node: prism::InterpolatedStringNode,
+) {
+    if ps.at_start_of_line() {
+        ps.emit_indent();
+    }
+
+    let is_heredoc = interpolated_string_node
+        .opening_loc()
+        .map(|s| loc_to_string(s).starts_with("<"))
+        .unwrap_or(false);
+
+    // Prism actually handles string concatenation when using "\", so it treats
+    // ```ruby
+    // "foo" \
+    //   "bar"
+    // ```
+    // as an interpolated node with the contents `"foobar"` (in two `parts` of "foo" and "bar").
+    // To detect this, we can look for any `InterpolatedStringNode` that has multiple `parts`
+    // and isn't a heredoc.
+    let is_backslash_string_interpolation = !is_heredoc
+        && interpolated_string_node
+            .parts()
+            .iter()
+            .all(|node| node.as_string_node().unwrap().opening_loc().is_some());
```

Comment:
This `unwrap()` does not feel guaranteed to be safe, but I don't have an obvious counterexample available.  Something like `.map(|s| s.opening_loc).is_some()` would be safer?